### PR TITLE
update tree shaking guide for webpack 4

### PR DESCRIPTION
**Pull request by probablyup**
updated the treeshaking documentation for webpack 4

### CODE REVIEW

**Comment by probablyup** [comment]
cc @sokra 

I think uglify isn't technically needed anymore to do tree shaking right? But seems like a good recommendation to enable it anyway.

**Comment by jeremenichelli** [comment]
Thanks @probablyup, can you change the base from `master`to `next` we are mergin everything there first. Thanks!

**Comment by probablyup** [comment]
@jeremenichelli done

File: src/content/guides/getting-started.md [hunk 1]
```diff
@@ -144,7 +144,7 @@ In this setup, `index.js` explicitly requires `lodash` to be present, and binds
 With that said, let's run `npx webpack` with our script as the [entry point](/concepts/entry-points) and `bundle.js` as the [output](/concepts/output). The `npx` command, which ships with Node 8.2 or higher, runs the webpack binary (`./node_modules/.bin/webpack`) of the webpack package we installed in the beginning:
 
 ``` bash
-npx webpack src/index.js dist/bundle.js
+npx webpack src/index.js --output dist/bundle.js
```
Comment (probablyup) [inline src/content/guides/getting-started.md]: not sure where this change came from...

File: src/content/guides/tree-shaking.md [hunk 2]
```diff
@@ -107,23 +108,46 @@ function cube(x) {
 Note the `unused harmony export square` comment above. If you look at the code below it, you'll notice that `square` is not being imported, however, it is still included in the bundle. We'll fix that in the next section.
 
 
-## Minify the Output
+## Mark the file as side-effect-free
+
+In a 100% ESM module world, identifying side effects is straightforward. However, we don't live in that reality just yet, so it's necessary to provide hints to webpack's compiler on the "pureness" of your code.
```
Comment (jeremenichelli) [inline src/content/guides/tree-shaking.md]: Can you rephrase `we don't live in that reality just yet, so` to `we aren't there just yet, in the mean time`

File: src/content/guides/tree-shaking.md [hunk 3]
```diff
@@ -107,23 +108,46 @@ function cube(x) {
...
+  "name": "your-project",
+  "sideEffects": false
+}
+```
 
-So we've cued up our "dead code" to be dropped by using the `import` and `export` syntax, but we still need to drop it from the bundle. To do that, we'll add a minifier that supports dead code removal -- the [`UglifyJSPlugin`](/plugins/uglifyjs-webpack-plugin) -- to our configuration...
+All the code noted above does not contain side effects, so we can simply mark the property as `false` to inform webpack that it can safely prune unused exports.
 
-Let's start by installing it:
+T> A "side effect" is defined as code that performs a special behavior when imported, other than simply exposing one or more exports. For instance: a polyfill which adjusts the global scope but does not provide an export or has one that you are not using in your codebase.
```
Comment (jeremenichelli) [inline src/content/guides/tree-shaking.md]: Can we rephrase this as:

```
A "side effect" is defined as code that performs a special behavior when imported, other than exposing one or more exports like a polyfill, which might affect the global scope and might not provide an export or has one that you are not using in your codebase.
```

File: src/content/guides/tree-shaking.md [hunk 4]
```diff
@@ -107,23 +108,46 @@ function cube(x) {
...
+    "./src/some-side-effectful-file.js"
+  ]
+}
 ```
 
-And then adding it into our config:
+## Minify the Output
+
+So we've cued up our "dead code" to be dropped by using the `import` and `export` syntax, but we still need to drop it from the bundle. To do that, we'll use the `-p` (production) webpack compilation flag to enable the uglifyjs minification plugin.
+
```
Comment (jeremenichelli) [inline src/content/guides/tree-shaking.md]: @TheLarkInn @sokra is the -p flag still recommended in webpack 4? Or should we clarify that is for versions older than v4?

File: src/content/guides/tree-shaking.md [hunk 5]
```diff
@@ -132,57 +156,21 @@ module.exports = {
...
-
-The code used in this guide assumes you perform tree-shaking using UglifyJS plugin. However, there are other tools such as [webpack-rollup-loader](https://github.com/erikdesjardins/webpack-rollup-loader) that may produce different results depending on your setup.
-
-
 ## Conclusion
 
 So, what we've learned is that in order to take advantage of _tree shaking_, you must...
 
 - Use ES2015 module syntax (i.e. `import` and `export`).
+- Add a "sideEffects" entry to your project's `package.json` file.
```
Comment (jeremenichelli) [inline src/content/guides/tree-shaking.md]: This would be _if_ the code is not affecting outside scope, not _always_ so we should add it to here.

Comment (probablyup) [reply src/content/guides/tree-shaking.md]: Actually no, you should always add a sideEffects entry or treeshaking won't work.

Comment (jeremenichelli) [reply src/content/guides/tree-shaking.md]: Well, but if your librari does produce side effects you shouldn't, right? I don't want to give that message to package owners.

Comment (probablyup) [reply src/content/guides/tree-shaking.md]: Perhaps, but in that case set it to `true`, or provide an array with the specific files that include side effects as described in the guide.

It's better to explicitly have the sideEffects known to webpack than to try and infer it.

Comment (montogeek) [reply src/content/guides/tree-shaking.md]: It is required?

Comment (jeremenichelli) [reply src/content/guides/tree-shaking.md]: Maybe is redundant since you already define what `sideEffects` means. I'm OK with it. Sorry for the noise.

File: src/content/guides/getting-started.md (hunk 1 above)
Comment (jeremenichelli) [reply src/content/guides/getting-started.md]: From the `next` rebase maybe.

**Comment by jeremenichelli** [comment]
There are some rephrasing that I think is necessary, the rest looks good @probablyup

**Comment by probablyup** [comment]
@jeremenichelli updated

**Comment by edmorley** [comment]
Do you think it's worth also linking to the [debugging optimization bailouts](https://webpack.js.org/plugins/module-concatenation-plugin/#debugging-optimization-bailouts) section from this page?

**Comment by edmorley** [comment]
I think this should also include:
* the new `sideEffects` selectors added in webpack/webpack/pull/6536
* setting `sideEffects` via module rules (ie allowing downstream apps to override the package.json entry) - webpack/webpack/pull/6129

**Comment by probablyup** [comment]
@edmorley updated

**Comment by probablyup** [comment]
@jeremenichelli could this be merged? I think many people would find it helpful

**Comment by edmorley** [comment]
@probablyup this is great - thank you for the changes :-)
